File: app/src/main/java/com/ryanrvldo/moviecatalogue/viewmodel/SearchViewModel.java
package com.ryanrvldo.moviecatalogue.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ryanrvldo.moviecatalogue.data.Repository;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.MovieResponse;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.TvShowResponse;

public class SearchViewModel extends ViewModel {

    private LiveData<MovieResponse> searchMovies;
    private LiveData<TvShowResponse> searchTv;

    private Repository repository;

    public SearchViewModel(Repository mRepository) {
        this.repository = mRepository;
    }

    public void setSearchMovies(String query) {
        searchMovies = new MutableLiveData<>();
        searchMovies = repository.searchMovies(query);
        Log.d("SearchMovie", "Fetched");
    }

    public LiveData<MovieResponse> getSearchMovies(String query) {
        if (searchMovies == null) {
            setSearchMovies(query);
        }
        return searchMovies;
    }

    public void setSearchTv(String query) {
        searchTv = new MutableLiveData<>();
        searchTv = repository.searchTvShows(query);
        Log.d("SearchTv", "DataFetched");
    }

    public LiveData<TvShowResponse> getSearchTv(String query) {
        if (searchTv == null) {
            setSearchTv(query);
        }
        return searchTv;
    }
}

File: app/src/main/java/com/ryanrvldo/moviecatalogue/ui/viewmodel/SearchViewModel.kt
package com.ryanrvldo.moviecatalogue.ui.viewmodel

import androidx.hilt.Assisted
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.ryanrvldo.moviecatalogue.data.Repository
import com.ryanrvldo.moviecatalogue.data.remote.response.MovieResponse
import com.ryanrvldo.moviecatalogue.data.remote.response.TvShowResponse
import com.ryanrvldo.moviecatalogue.data.vo.Resource

class SearchViewModel @ViewModelInject constructor(
    private val repository: Repository,
    @Assisted private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    private var query = savedStateHandle[SEARCH_QUERY_KEY] ?: ""

    val searchMovies: LiveData<Resource<MovieResponse>> by lazy {
        repository.searchMovies(query)
    }

    val searchTvShows: LiveData<Resource<TvShowResponse>> by lazy {
        repository.searchTvShows(query)
    }

    companion object {
        const val TAG = "SearchVM"
        const val SEARCH_QUERY_KEY = "SEARCH_QUERY"
    }
}